### ISSUE TITLE
SPAs and environment variables

### ISSUE THREAD

**Issue by mwcz** [spaship/spaship#77]
@Jared-Sprague was telling me about a conversation with @kunyan about environment variables and SPAs.  Just opened this issue so that @kunyan and @Jared-Sprague can fill in the details.

**Comment by kunyan** [comment]
Each component deployed on OpenShift will hosts as a Service.
It will set some environment variables into all exists Pods.

it will like
```shell

HTTPD_SERVICE_HOST=172.30.39.28
HTTPD_SERVICE_PORT=8080
PATH_PROXY_SERVICE_HOST=172.30.43.185
PATH_PROXY_SERVICE_PORT=8080
SYNC_SERVICE_SERVICE_HOST=172.30.7.32
SYNC_SERVICE_SERVICE_PORT=8080
```

We should set those env vars into each component config. then it will be connected automatically
And OpenShift user also could overwrite it to other hosts or services

**Comment by mwcz** [comment]
Ah, thanks for the clarification @kunyan.  From my conversation with @Jared-Sprague I thought there was talk of making certain environment variables available to the front-end.  Was that part of your thinking as well?

I was thinking it could be useful in scenarios like the following.

On the host running spa-manager, we could make a var `SPASHIP_SYNC_SERVICE_HOST` which would tell the spa-manager UI which sync-service to connect to.  That var would be useful in the backend and the frontend.  We could inject it into the frontend, perhaps like:

```html
<!doctype html>
<html>
  <head>
    <script>
      window.env = {
        SPASHIP_SYNC_SERVICE_HOST: "172.30.7.32"
      };
    </script>
  </head>
  <body>
    SPA MANAGER!
  </body>
</html>
```

However, I'll admit to knowing very little about OpenShift best practices.  Perhaps services & routes configuration can be used to handle the problem of configuring which API hosts a front-end should connect to.

**Comment by kunyan** [comment]
I think the SPA-Manager should only deploy one instance. and it should be able to connect all environments.


Here has a Download UI manager example: https://ziahamza.github.io/webui-aria2/
It just host on Github Page, and could service to localhost server.

I'm thinking about if SPA-Manager could cross over any project, any environments.
User could just put their api config into the UI. it will work

**Comment by mwcz** [comment]
I agree to an extent.  SPA Manager should be able to connect to multiple SPAship instances (dev, qa, stage, etc), but not _all_ SPAship instances.

I don't think we should be asking users to plug URLs into SPA Manager before they can use it.

I think that SPA Manager should be pre-configured with a list of SPAship APIs that it can communicate with.  When we deploy SPAship for a site, catalog for example, catalog should have one SPA Manager instance and four SPAship API instances (dev, qa, stage, prod).  That way, when a user signs in to the Catalog SPAship, they'll be able to deploy to dev/qa/stage/prod immediately without having to look up four URLs and follow an error-prone process of pasting them into SPA Manager.  Also, when we update Catalog's SPAship, the API and UI will be updated in tandem.

It's easier for users to understand, and it avoids issues like versioning mismatch where the UI and API are at different versions.

**Comment by kunyan** [comment]
This issue should be resolved by #167
